#114 caret position variable

vrajur opened the pull request:
# Description
Adds a new variable `caret_position` to the list of supported variables for the shell command plugin. This PR is current with development in the `0.8.0` branch. 

Change Type: New Feature

See related discussion in #114

Taitava commented on src/variables/ShellCommandVariable_CaretPosition.ts:
```diff
@@ -0,0 +1,98 @@
+
+export class ShellCommandVariable_CaretPosition extends ShellCommandVariable {
+    static variable_name = "caret_position";
+    static help_text = "Gives the line number and column position of the current caret position as 'line:col'. Get only the line number using {{caret_position:line}}, and only the column with {{caret_position:column}}. Line and column numbers are 1-indexed.";
```
Small change: 'line:col' to 'line:column'.
```suggestion
    static help_text = "Gives the line number and column position of the current caret position as 'line:column'. Get only the line number using {{caret_position:line}}, and only the column with {{caret_position:column}}. Line and column numbers are 1-indexed.";
```

Taitava commented on src/variables/ShellCommandVariable_CaretPosition.ts:
```diff
@@ -0,0 +1,98 @@
+
+    generateValue(): string {
+        // Check that we are able to get an editor
+        let editor = getEditor(this.app);
```
I think the variable will not be changed, so it can be a constant?

```suggestion
        const editor = getEditor(this.app);
```

Taitava commented on src/variables/ShellCommandVariable_CaretPosition.ts:
```diff
@@ -0,0 +1,98 @@
+            return null;
+        }
+
+        let position = editor.getCursor('to');
```
I think the variable will not be changed, so it can be a constant?
```suggestion
        const position = editor.getCursor('to');
```

Taitava commented on src/variables/ShellCommandVariable_CaretPosition.ts:
```diff
@@ -0,0 +1,98 @@
+
+        let position = editor.getCursor('to');
+        const line = position.line + 1; // editor position is zero-indexed, line numbers are 1-indexed
+        const col = position.ch + 1; // editor position is zero-indexed, column positions are 1-indexed
```
I like to use long variable names rather than abbreviations as I think they are easier and quicker to read and understand. Can you please change `col` to `column`? 🙂

vrajur replied:
Sure thing :slightly_smiling_face:

Taitava commented on src/variables/ShellCommandVariable_CaretPosition.ts:
```diff
@@ -0,0 +1,98 @@
+                case "column":
+                    return `${col}`;
+                default:
+                    Error("ShellCommandVariable_CaretPosition: Unrecognised argument: "+this.arguments.mode);
```
Variables have their own error handler.
```suggestion
                    this.newErrorMessage("Unrecognised argument: "+this.arguments.mode);
```

This shows the error message to a user in various ways: if executing a command, it will be shown in a notification balloon; if writing commands in settings, the preview text will show the error message. Variable name is automatically added to the message.

Taitava commented on src/variables/ShellCommandVariable_CaretPosition.ts:
```diff
@@ -0,0 +1,98 @@
+                    return `${col}`;
+                default:
+                    Error("ShellCommandVariable_CaretPosition: Unrecognised argument: "+this.arguments.mode);
+                    break;
```
```suggestion
                    return null;
```

Just to make this consistent to what other `{{variables}}` return when they encounter errors.

Taitava commented on src/variables/ShellCommandVariable_CaretPosition.ts:
```diff
@@ -0,0 +1,98 @@
+            // Normal variables
+            <IAutocompleteItem>{
+                value: "{{" + this.variable_name + "}}",
+                help_text: "Gives the line number and column position of the current caret position as 'line:col'.",
```
```suggestion
                help_text: "Gives the line number and column position of the current caret position as 'line:column'.",
```

Taitava commented on src/variables/ShellCommandVariable_CaretPosition.ts:
```diff
@@ -0,0 +1,98 @@
+                value: "{{!" + this.variable_name + "}}",
+                help_text: "Gives the line number and column position of the current caret position as 'line:col'.",
+                group: "Variables",
+                type: "normal-variable"
```
```suggestion
                type: "unescaped-variable",
```

I've forgotten myself to add the `type`s in `ShellCommandVariable_FolderPath` and `ShellCommandVariable_FilePath`. My bad!

Taitava commented on src/variables/ShellCommandVariable_CaretPosition.ts:
```diff
@@ -0,0 +1,98 @@
+                value: "{{!" + this.variable_name + ":line}}",
+                help_text: "Gives the line number of the current caret position.",
+                group: "Variables",
+                type: "normal-variable"
```
```suggestion
                type: "unescaped-variable",
```

Taitava commented on src/variables/ShellCommandVariable_CaretPosition.ts:
```diff
@@ -0,0 +1,98 @@
+                value: "{{!" + this.variable_name + ":column}}",
+                help_text: "Gives the column number of the current caret position.",
+                group: "Variables",
+                type: "normal-variable"
```
```suggestion
                type: "unescaped-variable",
```

Taitava commented on src/variables/ShellCommandVariable_CaretPosition.ts:
```diff
@@ -0,0 +1,98 @@
+}
+
+addShellCommandVariableInstructions(
+    "{{caret_position}} or {{caret_position:line}} or {{caret_position:column}}",
```
```suggestion
    "{{caret_position}}, {{caret_position:line}} or {{caret_position:column}}",
```

Taitava commented on src/variables/ShellCommandVariable_CaretPosition.ts:
```diff
@@ -0,0 +1,98 @@
+            // Unescaped variables
+            <IAutocompleteItem>{
+                value: "{{!" + this.variable_name + "}}",
+                help_text: "Gives the line number and column position of the current caret position as 'line:col'.",
```
```suggestion
                help_text: "Gives the line number and column position of the current caret position as 'line:column'.",
```

Taitava reviewed (changes requested):
Please also include issue number in the commit messages. Personally, I like this style, where the commit number is the first thing: `#119: Commit comment`. It's an easy reference for understanding where the commit relates to.

I actually forgot to create an issue at first, but now it exists (#119) and you can e.g. squash commit your PR to have the issue number prefixed in a commit message. Normally I do not expect squash committing, as it can lose details of history, but this PR is not huge, and squashing feels a lot easier way to add the issue number to commit messages, than what it would be to edit old commit messages.

EDIT: And thank you very much for your work! 🙂 I think this is the first pull request I've ever had on any of my public projects! 😎

Taitava commented:
# I created an AUTHORS.md file in the `0.8.0` branch
https://github.com/Taitava/obsidian-shellcommands/blob/0.8.0/AUTHORS.md

If you want, you can add your information there. Just include the commit in this same PR. 🙂

vrajur commented:
Thanks for the review! I updated all your suggestions, updated the AUTHORS.md file and squashed the changes into a single commit (formatted as you specified). Let me know if there's anything else. 

I also opened an associated [PR in the docs repository](https://github.com/Taitava/obsidian-shellcommands-documentation/pull/1) with a very simple example for you to take a look at as well. 

Thanks!

Taitava commented on AUTHORS.md:
```diff
@@ -11,6 +11,16 @@ Contact:
 
+
+## Vinay Rajur
+ - Contributor
```
I forgot to write to the instructions that you can also mention what you have done, e.g. _Created new variable `{{caret_position}}` #114 / #119_. But I can add this mention next time I update the file.

vrajur replied:
No problem, in that case feel free to go ahead and update it :slightly_smiling_face:

Taitava reviewed (commented):
Thank you, all ok now! 🙂 I'll approve merging when I have the `0.9.0` branch created after `0.8.0` is released.

Edit: Actually I'll create the branch already now.

Taitava commented:
# Changed the license from MIT to GPL-3.0

Hi @vrajur ! 🙂 

I've decided to change the license of this project to GPL. At the same time, I added copyright headers to all source code files, including the one you made. Can you check the copyright statement here and tell me if it's okay to you: https://github.com/Taitava/obsidian-shellcommands/blob/5e155e960cea25717d5575b4523b19cec3f9f3c2/src/variables/Variable_CaretPosition.ts

The updated version of `Variable_CaretPosition.ts` that contains the copyright statement is not yet in the `main` branch, it's still in `#199-change-license-to-GPL`, but will be merged to `main` when I release `0.12.0`.

Thank you again for your great contribution! 😎
